src/main/java/io/wispforest/owo/serialization/Codeck.java

```diff
@@ -0,0 +1,306 @@
+            if(deserializer instanceof SelfDescribedDeserializer<E> selfDescribedDeserializer){
+                var nbtSerializerzer = new NbtSerializer();
+
+                nbtSerializerzer.writeAny(selfDescribedDeserializer.readAny());
+
+                return nbtSerializerzer.result();
+            }
+
+            byte[] array = Codeck.BYTE_ARRAY.decode(deserializer);
+
+            try {
+                ByteArrayDataInput stream = ByteStreams.newDataInput(array);
+
+                return NbtIo.read(stream, NbtTagSizeTracker.ofUnlimitedBytes());
+            } catch (IOException e){
+                LOGGER.error("Unable to deserialize the given format into the desired NbtElement!");
+                throw new RuntimeException(e);
+            }
+        }
+    };
+
+    Codeck<NbtCompound> COMPOUND = Codeck.NBT_ELEMENT.then(element -> ((NbtCompound) element), compound -> compound);
+
+    //--
+
+    Codeck<Identifier> IDENTIFIER = Codeck.STRING.then(Identifier::new, Identifier::toString);
+
+    Codeck<ItemStack> ITEM_STACK = COMPOUND
+            .then(ItemStack::fromNbt, stack -> stack.writeNbt(new NbtCompound()));
+
+    Codeck<UUID> UUID = StructCodeckBuilder.of(
+            StructField.of("mostSig", Codeck.LONG, java.util.UUID::getMostSignificantBits),
+            StructField.of("leastSig", Codeck.LONG, java.util.UUID::getLeastSignificantBits),
+            UUID::new
+    );
+
+    Codeck<Date> DATE = Codeck.LONG.then(Date::new, Date::getTime);
+
+    Codeck<PacketByteBuf> PACKET_BYTE_BUF = Codeck.BYTE_ARRAY
+            .then(
+                    bytes -> {
+                        var byteBuf = PacketByteBufs.create();
+
+                        byteBuf.writeBytes(bytes);
+
+                        return byteBuf;
+                    },
+                    byteBuf -> {
+                        var bytes = new byte[byteBuf.readerIndex()];
+
+                        byteBuf.readBytes(bytes);
+
+                        return bytes;
+                    }
+            );
+
+    Codeck<BlockPos> BLOCK_POS = Codeck.LONG.then(BlockPos::fromLong, BlockPos::asLong);
+    Codeck<ChunkPos> CHUNK_POS = Codeck.LONG.then(ChunkPos::new, ChunkPos::toLong);
+
+    Codeck<BitSet> BITSET = Codeck.LONG_ARRAY.then(BitSet::valueOf, BitSet::toLongArray);
+
+    Codeck<Text> TEXT = Codeck.STRING.then(Text.Serializer::fromJson, Text.Serializer::toJson);
+
+
+    //--
+
+    //Kinda mega cursed but...
+    static <e, T> Codeck<T> of(BiConsumer<Serializer<e>, T> encode, Function<Deserializer<e>, T> decode) {
+        return new Codeck<T>() {
+            @Override
+            public <E> void encode(Serializer<E> serializer, T value) {
+                encode.accept((Serializer<e>) serializer, value);
+            }
+
+            @Override
+            public <E> T decode(Deserializer<E> deserializer) {
+                return decode.apply((Deserializer<e>) deserializer);
+            }
+        };
+    }
```

Comment:
This would be a bit cleaner if `encode` and `decode` were bespoke interfaces that had the proper generics in them

Reply:
I really would not want to split such into separate interfaces